python/ray/util/sgd/pytorch/examples/dcgan.py

```diff
@@ -114,119 +95,137 @@ def forward(self, x):
-            pyx = part[i, :]
-            scores.append(entropy(pyx, py))
-        split_scores.append(np.exp(np.mean(scores)))
-
-    return np.mean(split_scores), np.std(split_scores)
+def model_creator(config):
+    def weights_init(m):
+        classname = m.__class__.__name__
+        if classname.find("Conv") != -1:
+            nn.init.normal_(m.weight.data, 0.0, 0.02)
+        elif classname.find("BatchNorm") != -1:
+            nn.init.normal_(m.weight.data, 1.0, 0.02)
+            nn.init.constant_(m.bias.data, 0)
 
+    discriminator = Discriminator()
+    discriminator.apply(weights_init)
 
-def model_creator(config):
-    netD = Discriminator()
-    netD.apply(weights_init)
-
-    netG = Generator()
-    netG.apply(weights_init)
-    return netD, netG
-
-
-def train(config, models, dataloader, criterion, optimizers, **kwargs):
-    netD, netG = models
-    optimD, optimG = optimizers
-    real_label = 1
-    fake_label = 0
-    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
-
-    for i, data in enumerate(dataloader, 0):
-        if i >= TRAIN_BATCHES and config.get(TEST_MODE):
-            break
-
-        netD.zero_grad()
-        real_cpu = data[0].to(device)
-        b_size = real_cpu.size(0)
-        label = torch.full((b_size, ), real_label, device=device)
-        output = netD(real_cpu).view(-1)
-        errD_real = criterion(output, label)
+    generator = Generator()
+    generator.apply(weights_init)
+    return discriminator, generator
+
+
+def optimizer_creator(models, config):
+    net_d, net_g = models
+    discriminator_opt = optim.Adam(
+        net_d.parameters(), lr=config.get("lr", 0.01), betas=(0.5, 0.999))
+    generator_opt = optim.Adam(
+        net_g.parameters(), lr=config.get("lr", 0.01), betas=(0.5, 0.999))
+    return discriminator_opt, generator_opt
+
+
+class GANOperator(TrainingOperator):
+    def setup(self, config):
+        self.device = torch.device("cuda"
+                                   if torch.cuda.is_available() else "cpu")
+
+        self.classifier = LeNet()
+        self.classifier.load_state_dict(
+            torch.load(config["classification_model_path"]))
+        self.classifier.eval()
+
+    def inception_score(self, imgs, batch_size=32, splits=1):
+        N = len(imgs)
+        dataloader = torch.utils.data.DataLoader(imgs, batch_size=batch_size)
+        up = nn.Upsample(
+            size=(28, 28), mode="bilinear").type(torch.FloatTensor)
+
+        def get_pred(x):
+            x = up(x)
+            x = self.classifier(x)
+            return F.softmax(x).data.cpu().numpy()
+
+        preds = np.zeros((N, 10))
+        for i, batch in enumerate(dataloader, 0):
+            batch = batch.type(torch.FloatTensor)
+            batchv = Variable(batch)
+            batch_size_i = batch.size()[0]
+            preds[i * batch_size:i * batch_size +
+                  batch_size_i] = get_pred(batchv)
+
+        # Now compute the mean kl-div
```

Comment:
This looks weird without any preceding comments. Maybe add some more similar ones for other blocks?

Reply:
Done